### PR TITLE
Reduce the size of the binary and image size by upx

### PR DESCRIPTION
**Pull request by LinuxSuRen**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
[upx](https://github.com/upx/upx) is a tool that can reduce the binary size dramatically.

You can see the reduce ratio below:
```
        File size         Ratio      Format      Name
   --------------------   ------   -----------   -----------
 108234000 ->  41676816   38.51%   macho/amd64   ks-apiserver 


         File size         Ratio      Format      Name
   --------------------   ------   -----------   -----------
  72386928 ->  28536848   39.42%   macho/amd64   controller-manager
```

Before we using upx, the docker images size is:
```
surenpi/ks-controller-manager   reduce-binary-size                               262ebb9fa9d4       78.5MB
surenpi/ks-apiserver            reduce-binary-size                               90f4774c0370       114MB
```

After we using upx, the docker images size is:
```
surenpi/ks-controller-manager   reduce-binary-size                               dc649966e3ec       34.7MB
surenpi/ks-apiserver            reduce-binary-size                               3b866f1b8f44       47.8MB
```

**Which issue(s) this PR fixes**:
None.

**Special notes for reviewers**:

**Additional documentation, usage docs, etc.**:

Want to get started on your own computer? It's very easy to install for macOS users. For other users, you can download it from the [release page](https://github.com/upx/upx/releases).

```
brew install upx
```

### CODE REVIEW

**Comment by RolandMa1986** [comment]
Seems we should also add **-ldflags="-s -w"** flag to strip the binary output.

**Comment by LinuxSuRen** [comment]
@RolandMa1986 Sure. I on your side with these flags.

Before using flags `-w -s`:
```
-rwxr-xr-x  1 rick  staff    40M Dec  1 16:21 bin/cmd/ks-apiserver
UPX 3.96        Markus Oberhumer, Laszlo Molnar & John Reiser   Jan 23rd 2020

        File size         Ratio      Format      Name
   --------------------   ------   -----------   -----------
 108234000 ->  41676816   38.51%   macho/amd64   ks-apiserver
```

After using flags `-w -s`:
```
-rwxr-xr-x  1 rick  staff    22M Dec  1 16:26 bin/cmd/ks-apiserver
UPX 3.96        Markus Oberhumer, Laszlo Molnar & John Reiser   Jan 23rd 2020

        File size         Ratio      Format      Name
   --------------------   ------   -----------   -----------
  89310480 ->  22753296   25.48%   macho/amd64   ks-apiserver
```

**Comment by LinuxSuRen** [comment]
/cc @RolandMa1986
/cc @zryfish 

Any comments here?

**Comment by RolandMa1986** [comment]
Using upx seems a little bit overdone here.  Only disk spaces are saved by compressing the binary, But it's not critical for most users. And it won't save network bandwidth, since the Image will be compressed during transport(a compressed file can hardly be shrunk again).

File: go.mod [hunk 1]
```diff
@@ -51,6 +51,7 @@ require (
 	github.com/kubernetes-csi/external-snapshotter/v2 v2.1.0
 	github.com/kubesphere/sonargo v0.0.2
 	github.com/lib/pq v1.2.0 // indirect
+	github.com/mitchellh/mapstructure v1.2.2
```
Comment (shaowenchen) [inline go.mod:54]: Why do we need to add a new library? It seems not being used.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -39,7 +39,11 @@ jobs:
         run: bash hack/install_kubebuilder.sh
 
       - name: Build
-        run: make all
+        run: |
+          export PATH=$PATH:${PWD}/upx-3.96-amd64_linux
+          curl -L https://github.com/upx/upx/releases/download/v3.96/upx-3.96-amd64_linux.tar.xz | tar xJ
+          upx -V
+          make all
```
Comment (shaowenchen) [inline .github/workflows/build.yml:46]: Why do we not modify nightly-builds？We should modify every builder, not only this.

**Comment by shaowenchen** [comment]
It is necessary to provide comparative data between the two cases.

**Comment by LinuxSuRen** [comment]
> It is necessary to provide comparative data between the two cases.

I still need to find a way to get the difference when transferring the images.

**Comment by shaowenchen** [comment]
There is a conflict that needs to be resolved.

**Comment by zryfish** [comment]
UPX is considered malicious by some Security Scan software, and there are some complaints from users, we'd better keep the executables the way it is.

**Comment by LinuxSuRen** [comment]
> UPX is considered malicious by some Security Scan software, and there are some complaints from users, we'd better keep the executables the way it is.

Fair enough. I'm going to close this PR until we have a better reason.

/close

**Pull request closed without merging**